File: Data Structures & Algorithms/maximum-subarray/submission-5.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # s = 0
        # n = len(nums)
        # for i in range(n):
        #     s = nums[i]
        #     for j in range(i + 1, n):
        #         s = max(s, s + nums[j])
        #         print(s)
        # return s
        s = 0
        l = 0
        res = nums[0]
        for num in nums:
            if num < 0:
                l += 1
        #     else:
        #         s += num
        #         res = num
        #         break
        if l == len(nums):
            return max(nums)
        for r in range(len(nums)):
            s += nums[r]
            if s < 0:
                s = 0
            res = max(res, s)
            print(res)
        return res
            
            


        

File: Data Structures & Algorithms/three-integer-sum/submission-5.rs
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort();
        let mut res = Vec::new();
        let n = nums.len();

        for i in 0..n {
            let a = nums[i];
            if a > 0 {
                break;
            }
            if i > 0 && a == nums[i - 1] {
                continue;
            }

            let (mut l, mut r) = (i + 1, n - 1);
            while l < r {
                let sum = a + nums[l] + nums[r];
                if sum > 0 {
                    r -= 1;
                } else if sum < 0 {
                    l += 1;
                } else {
                    res.push(vec![a, nums[l], nums[r]]);
                    l += 1;
                    r -= 1;
                    while l < r && nums[l] == nums[l - 1] {
                        l += 1;
                    }
                }
            }
        }
        res
    }
}
